DO NOT MERGE: trying out setuptools PR

henryiii opened the pull request:
<!--
Title (above): please place [branch_name] at the beginning if you are targeting a branch other than master. *Do not target stable*.
It is recommended to use conventional commit format, see conventionalcommits.org, but not required.
-->
## Description

<!-- Include relevant issues or PRs here, describe what changed and why -->


This is based on #3719.

See https://github.com/henryiii/pybind11/pull/9 too.

henryiii committed:
DO NOT MERGE: trying out setuptools PR

abravalheri committed:
DO NOT MERGE: Update git ref for experimental setuptools branch

abravalheri committed:
DO NOT MERGE: Add project metadata to pyproject.toml

henryiii committed:
DO NOT MERGE: drop Python 3.6 from the CI

henryiii commented:
So far:

```
E  'pybind11_global.egg-info/not-zip-safe',
```
is missing with the TOML config.

abravalheri commented:
Thanks @henryiii, I will check!

henryiii commented:
Okay, that's weird, a single job has this:

```
AttributeError: module 'pybind11' has no attribute 'get_include'
```

I think it's related to the not-zip-safe issue, at least that's my guess. That's from `pybind11-global`, which is the weird legacy version that has "global" data and headers, rather than being inside a Python package.

abravalheri commented:
Ok, I tried the following:

```bash
% PYBIND11_GLOBAL_SDIST=1 .python -m buildls dist
% ls dist
pybind11-2.10.0.dev1.tar.gz  pybind11_global-2.10.0.dev1-py3-none-any.whl
```

and I noticed the names of the files produced... The `sdist` name does not contain `_global`...

I think this is caused by the fact that PEP 621 requires `name` to be statically defined in the `pyproject.toml` (so in theory trying to redefine it dynamically should not be an option).

The `tar.gz` file however does contain a `not-zip-safe` member:

```bash
% tar tf dist/pybind11-2.10.0.dev1.tar.gz | grep zip
pybind11-2.10.0.dev1/pybind11.egg-info/not-zip-safe
```

abravalheri commented:
@henryiii, I think that for `pybind11_global` you will have to edit `pyproject.toml` before running the build 😅

abravalheri commented:
> ```
> E         -  'pybind11_global.egg-info/not-zip-safe',
> ```
> 
> is missing with the TOML config.

I added a test case for that in https://github.com/pypa/setuptools/pull/3068/commits/ed73bc52fde8d3106295f24b9a45dea8292374bd. Let's see what happens with the test suite, but running the tests locally the `not-zip-safe` file seems to be generated.

henryiii commented:
Ah, yes, I wasn't really expecting this to fully work with PEP 621 metadata, I was originally seeing if it would still work with the changes and the old system. The static name is correct. I'd be perfectly fine adding some sort of pre-processing system to make the global SDist.

abravalheri committed:
DO NOT MERGE: Use 'build' in tests instead of running setup.py

It also temporarily renames the project for global builds

abravalheri committed:
DO NOT MERGE: Temporarily edit pyproject.toml for global build

abravalheri committed:
Ensure global package is built correctly on test_files

abravalheri committed:
Update setuptools branch for experimental installation

abravalheri committed:
Allow user to specify flags (e.g. -vv) to nox' test_packaging session

abravalheri committed:
Minor changes just to trigger a build

The experimental branch in setuptools got updated. It would be nice to
see if the checks still pass.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

abravalheri committed:
Adapt to changes in the handling of license for PEP 621

As discussed in
https://discuss.python.org/t/help-testing-experimental-features-in-setuptools/13821/59
the handling of `license` has changed to better complain with PEP 621.

The changes introduced here reflect that change.

abravalheri commented:
I updated the PR to reflect the changes discussed in https://discuss.python.org/t/help-testing-experimental-features-in-setuptools/13821/59, regarding
the handling of `license` to better complain with PEP 621.

abravalheri commented on pyproject.toml:
```diff
@@ -47,10 +48,6 @@ Chat = "https://gitter.im/pybind/Lobby"
 
-[tool.setuptools.dynamic]
-license = "BSD"
-license-files = ["LICEN[CS]E*", "COPYING*", "NOTICE*", "AUTHORS*"]
```
These are the default values already considered by `setuptools` and `wheel`.

Since it was clarified in the Python discourse that specifying `project.license` statically does not prevent the backend to dynamically populate `License-files` in the core metadata, having a dynamic `license` is no longer necessary.

abravalheri commented on pyproject.toml:
```diff
@@ -1,7 +1,53 @@
+Chat = "https://gitter.im/pybind/Lobby"
+
+[tool.setuptools]
+zip-safe = false
```
I wonder if this option is even necessary nowadays.

According to the docs: `it will consider the project unsafe if it contains any C extensions or datafiles whatsoever.`

But more importantly... This flag works for eggs right? I believe `pip` will always extract the package anyway...
